### ISSUE TITLE
Turn off -Werror or -Wimplicit-function-declaration by default for C code?

### ISSUE THREAD

**Issue by varungandhi-src** [rbenv/ruby-build#2001]
I ran into the OpenSSL build failure recently with `-Werror`. Out of curiosity, I looked at the issue tracker and it seems like the use of `-Werror` is a source of repeated build problems. Some examples:

* https://github.com/rbenv/ruby-build/issues/1998 due to -Werror + -Wimplicit-function-declaration in OpenSSL
* https://github.com/rbenv/ruby-build/issues/1747 due to -Werror + -Wimplicit-function-declaration in Ruby itself, I think
* https://github.com/rbenv/ruby-build/issues/1489 due to -Werror + -Wimplicit-function-declaration in OpenSSL
  * This had a dupe:  https://github.com/rbenv/ruby-build/issues/1764
* https://github.com/ffi/ffi/issues/869 due to -Werror + -Wimplicit-function-declaration in Ruby FFI
  * Dupe of this GitHub issue: https://github.com/rbenv/ruby-build/issues/1709#issuecomment-964526168 (converted to Discussion #1870)
  * Another dupe: https://github.com/rbenv/ruby-build/issues/1699 (converted to Discussion #1946)

Whenever something like this happens, it ends up wasting time for a bunch of people as evidenced by these threads, and it's not clear (at least, to me) what the benefit of having that error is if someone is going to recommend (the very reasonable) workaround of using `-Wno-error=implicit-function-declaration`. Would it make sense to change the default, by adding a  `-Wno-error` or `-Wno-error=implicit-function-declaration` prefix to the appropriate environment variables?

**Comment by eregon** [comment]
Yes, this is annoying but this `-Werror=implicit-function-declaration` is the default on macOS (and not on Linux, hence why it does not always get caught early). First, it does not seem advisable to disable something the OS does by default.

> if someone is going to recommend (the very reasonable) workaround of using `-Wno-error=implicit-function-declaration`.

Actually it can be a pretty dangerous workaround, because then the compiler does not know precise argument types and could end up passing arguments incorrectly, which could cause all kinds of problems (e.g., if the prototype is var-args it'd probably crash or corrupt the arguments on aarch64).
It is also always a clear bug in the source for not properly including functions it uses, and missing an include e.g. can result in incorrect compilation, not defining the macros as needed, not using an intrinsified version of the function, etc.

Also BTW recent CRuby enables that flag by default, and it would be very bad to disable something CRuby enables on purpose: https://bugs.ruby-lang.org/issues/18615
Since CRuby enables it for their CI and they test on macOS this should not happen anymore for CRuby.

Regarding OpenSSL, I'd wish they check their CI before releasing (they do enable the flag in CI as well) but apparently they did not for 1.1.1q.
